L7 test deletes security group and updated health check.

jtk54 opened the pull request:
The L7 test should not leak resources anymore. @ewiseblatt please review.

ewiseblatt commented:
I want to look into something before commenting. And also I'm going to use this to test out my leak detetector configuration. I should get to this by the end of the day.

jtk54 commented:
Ok, fine with me.

ewiseblatt commented:
Security group looks ok, but where is the health check created? Is that implicit when you changed it? How would the health check get reclaimed in practice?

jtk54 commented:
The health check is created new when we update the name of the health check in Upsert. I'm re-inserting it in the load balancer during the certificate tests, and then the final delete will remove all the components included in the load balancer. If it's not clear I can add more detailed comments or shift this around a bit.

jtk54 commented:
I updated the comments to be clearer, and added one into the actual test case enumeration. This works because we do the tests in a certain order, so I commented as such.

ewiseblatt commented on testing/citest/tests/google_http_lb_upsert_test.py:
```diff
@@ -168,6 +168,8 @@ def test_l_add_cert(self):
                                               'add cert'))
 
+  # Note: We include the updated health check in this upsert description.
+  # If the order of tests 'm' and 'o' change, we will leak resources.
```
name the tests since letters can change.

jtk54 commented:
The 'delete hc' is the 'delete load balancer' command. The load balancer delete deals with all the components in the load balancer, including health checks, backend services, forwarding rules, etc.

ewiseblatt commented on testing/citest/tests/google_http_lb_upsert_scenario.py:
```diff
@@ -480,7 +480,17 @@ def update_port_range(self):
     '''
+    # Note: We add this health check to the upsert description because
+    # this health check would be orphaned otherwise. This allows the 'delete'
+    # test to delete all the resources we allocated.
```
I'm still hung up on this.
If before the healthCheck was unchanged, so why would it be orphaned? Now you have a different one, so what happens to the original? If anything I'd think now would lead to something orphaned. Please explain.

jtk54 replied:
```python
  def test_c_upsert_min_lb(self): 
    self.run_test_case(self.scenario.upsert_min_load_balancer()) # => Use proto health check for all backend services.

  def test_d_upsert_full_lb(self):
    self.run_test_case(self.scenario.upsert_full_load_balancer()) # => Use proto health check for all backend services.

  def test_e_add_security_group(self):
    self.run_test_case(self.scenario.add_security_group())

  # Test letters f and n are reserved for another test derived from this one.

  def test_g_change_hc(self):
    self.run_test_case(self.scenario.change_health_check()) # => Create new, updated health check and use for all backend services. Old proto health check still exists in GCP and is orphaned.

  def test_h_change_bs(self):
    self.run_test_case(self.scenario.change_backend_service()) # => Re-use proto health check for all backend services. Now the updated health check is orphaned.

  def test_i_add_host_rule(self):
    self.run_test_case(self.scenario.add_host_rule()) # => Re-use proto health check for all backend services. Now the updated health check is orphaned.

  def test_j_update_host_rule(self):
    self.run_test_case(self.scenario.update_host_rule()) # => Re-use proto health check for all backend services. Now the updated health check is orphaned.

  def test_k_update_port_range(self):
    self.run_test_case(self.scenario.update_port_range()) # => Re-use proto health check for all backend services. Now the updated health check is orphaned.

  def test_l_add_cert(self):
    self.run_test_case(self.scenario.add_cert(self.__class__.FIRST_CERT, 'add cert')) # => Add the updated health check to just one of the backend services, so both the updated health check and the proto health check are included in the load balancer.

def test_m_change_cert(self):
    self.run_test_case(self.scenario.add_cert(self.__class__.SECOND_CERT,
                                              'update cert')) # => Add the updated health check to just one of the backend services, so both the updated health check and the proto health check are included in the load balancer.

  # Test letters f and n are reserved for another test derived from this one.

  def test_o_delete_lb(self):
    self.run_test_case(self.scenario.delete_http_load_balancer()) # => Delete all components currently associated with the load balancer.
 
```

In one of the upsert operations, we create a _new_ health check that is an 'updated' health check to the L7 load balancer. Because we replace _all_ instances of the old health check in the L7 load balancer, the old (and thus unused) health check is now orphaned in GCP.

We then switch back to using the original health check for all the health checks in the L7 for a few upserts, which then makes the 'updated' health check orphaned in GCP. Before this PR, we would no longer use the updated health check in any of the rest of the operations so it would leak. This PR now makes it so that we add the 'updated' health check back into the load balancer so we can delete it.

ewiseblatt commented:
My concern here is that I'd call this a spinnaker bug.
You shouldnt have to delete that. If spinnaker created it implicitly and it
is no longer needed, then spinnaker should delete it as well. Otherwise as
a normal user, you'd be leaking these as well.


On Tue, Oct 25, 2016 at 9:23 AM, Jacob Kiefer <notifications@github.com>
wrote:

> The health check is created new when we update the name of the health
> check in Upsert. I'm re-inserting it in the load balancer during the
> certificate tests, and then the final delete will remove all the components
> included in the load balancer. If it's not clear I can add more detailed
> comments or shift this around a bit.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spinnaker/spinnaker/pull/1262#issuecomment-256032852>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEKkGNve31xWa3bKCZwED95Rav5LVyEPks5q3gK5gaJpZM4Ke6Q9>
> .
>

jtk54 commented:
Ping -- I now delete the orphaned health check with an explicit API call in the the test cleanup function.
